gcc/rust/rust-session-manager.cc

```diff
@@ -1113,3 +1187,33 @@ TargetOptions::enable_implicit_feature_reqs (std::string feature)
  *  - code generation
  *  - link */
 } // namespace Rust
+
+#if CHECKING_P
+namespace selftest {
+void
+rust_crate_name_validation_test (void)
+{
+  ASSERT_TRUE (Rust::CompileOptions::validate_crate_name ("example"));
+  ASSERT_TRUE (Rust::CompileOptions::validate_crate_name ("abcdefg_1234"));
+  ASSERT_TRUE (Rust::CompileOptions::validate_crate_name ("1"));
+  // FIXME: The next test does not pass as of current implementation
+  // ASSERT_TRUE (Rust::CompileOptions::validate_crate_name ("惊吓"));
+  // NOTE: - is not allowed in the crate name ...
+  /*
+  ASSERT_FALSE (Rust::CompileOptions::validate_crate_name ("abcdefg-1234"));
+  ASSERT_FALSE (Rust::CompileOptions::validate_crate_name ("a+b"));
+  ASSERT_FALSE (Rust::CompileOptions::validate_crate_name ("/a+b/")); */
```

Comment:
I am not sure how to correctly handle the negative tests, since `error_at` itself will just fail the test.

Reply:
I'd say a way to do this is to have the `validate_crate_name` function signature look like this:
```c++
bool validate_crate_name(const std::string &name, Error &error);
```
If the input is invalid, return false and store the error inside the given reference. Then, the caller emits it
```c++
Error error;
if (!validate_crate_name(name, error)) {
    error.emit_error();
    return false;
}
```

Reply:
> I'd say a way to do this is to have the `validate_crate_name` function signature look like this:
> 
> ```c++
> bool validate_crate_name(const std::string &name, Error &error);
> ```
> 
> If the input is invalid, return false and store the error inside the given reference. Then, the caller emits it
> 
> ```c++
> Error error;
> if (!validate_crate_name(name, error)) {
>     error.emit_error();
>     return false;
> }
> ```

I think this is also valid. I am not sure which one is better? (1) clear the error flag somewhere in GCC; (2) pass the `Error` back to the caller.